### PR TITLE
feat(parser): Support parsing courses from a Github repository

### PR DESCRIPTION
**Pull request by arku**
This PR adds in support to parse courses from Github.

New options:
- token (the access token to read repositories)
- branch (the repository branch to pull contents from) - Corporate Training team maintains multiple courses in different branches.

You can run it with

```bash
curriculum-parser topic https://github.com/Laboratoria/mass-training \
--repo Laboratoria/mass-training \
--branch 05-Onboardingalicorp \
--track business \
--version 1.0.0 \
--locale es-ES \
--token <token> 
```

Please note that the solution is not perfect. We might want to rearchitecture the app to go for clean code.

### CODE REVIEW

File: lib/course.js [hunk 1]
```diff
@@ -263,12 +260,41 @@ const applyGlobalOptions = (result, opts, relativePath, parserVersion) => ({
...
+    const { data } = await octokit.repos.getContents({
+      owner: username,
+      repo: repository,
+      path: readmeFileName,
+      ref: branch,
+    });
+    readme = Buffer.from(data.content, 'base64').toString();
+    relativePath = dirOrPath;
+    // TODO: Check with the team if this is fine
+    slug = `${repository}-${branch}`;
```
Comment (arku) [inline lib/course.js:286]: @FabianBravoA I am using both the repository name and the branch name to make sure we don't have any duplicate slugs for courses from a single repository. We have courses maintained in branches in [`mass-training`](https://github.com/Laboratoria/mass-training), for example.

**Comment by arku** [comment]
@FabianBravoA There's a fair bit of repeated code in the files changed as I haven't found an abstraction yet. I just have code that works 🙈 . Open to discuss and come up with a better architecture.

We could, for example, let each of the individual functions from `syllabus`, `unit`, and `part` just accept the content as input and not know much information about the source it is coming from (File system or Github or any other exernal source).

With the current setup, there's a bit of coupling - the functions know where to look for the content and how to parse it. This goes against the [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single_responsibility_principle). We should instead delegate the course reading responsibility to some other function which will figure out how to get the data (from the file system or from Github). The content can then be passed to a function which will be responsible to parse it.

WDYT?

File: lib/course.js [hunk 2]
```diff
@@ -263,12 +260,41 @@ const applyGlobalOptions = (result, opts, relativePath, parserVersion) => ({
 //
 
 
-module.exports = (dir, { Topic }, opts = {}) => {
-  const readme = path.join(dir, common.getReadmeFileName(opts.locale));
-  const relativePath = path.relative(process.cwd(), dir);
-  const { slug } = common.parseDirname(readme);
+module.exports = async (dirOrPath, { Topic }, opts = {}) => {
```
Comment (arku) [inline lib/course.js:263]: The variable name `dirOrPath` is really ugly. Can you help me identify a better name?

File: lib/part.js [hunk 3]
```diff
@@ -34,30 +35,91 @@ const partFormats = {
...
+  duration: 'duration',
+  duração: 'duration',
+};
 
-module.exports = (dir, opts = {}) => {
+module.exports = async (dirOrPath, opts = {}) => {
   const readmeFileName = common.getReadmeFileName(opts.locale);
-  const readmeAbsPath = path.join(dir, readmeFileName);
 
+  if (opts.branch && opts.token) {
```
Comment (arku) [inline lib/part.js:51]: Is this fine to figure out if the content should be fetched from Github?

File: lib/course.js (hunk 2 above)
Comment (arku) [reply lib/course.js:263]: I was thinking of location which might refer to a location in the File System or a location on the web (a URL) but not sure if it is okay.

File: lib/course.js (hunk 1 above)
Comment (FabianBravoA) [reply lib/course.js:286]: As we said, we don't need a completely unique slug name, this is just fine for us now as we're uploading courses into cohorts directly

File: lib/part.js (hunk 3 above)
Comment (FabianBravoA) [reply lib/part.js:51]: Don't figure out this, we should always assume we're working with github. (Let's fork this completely into a new project from now on)

File: lib/course.js [hunk 4]
```diff
@@ -263,12 +260,41 @@ const applyGlobalOptions = (result, opts, relativePath, parserVersion) => ({
...
+  let readme;
+  let relativePath;
+  let slug;
+  let dir;
+  let location;
+
+  const readmeFileName = common.getReadmeFileName(opts.locale);
+  if (common.isGithubUrl(dirOrPath)) {
+    const { repo, token, branch } = opts;
+    const [username, repository] = repo.split('/');
```
Comment (FabianBravoA) [inline lib/course.js:273]: I don't know if we can get all this information from the parameters (ie: from the endpoint post data, for example) and not infer this from the URL.

File: lib/part.js [hunk 5]
```diff
@@ -34,30 +35,91 @@ const partFormats = {
...
+
+      if (parsedReadme.duration === null) {
+        throw Object.assign(new Error('Could not parse part duration'), {
+          path: dirOrPath,
+          type: 'part',
+        });
+      }
+
+      // TODO: Support parsing of `practice` type parts
+      // Not a priority for CT courses
```
Comment (FabianBravoA) [inline lib/part.js:82]: Let's forget this for now...

**Review by FabianBravoA — APPROVED** [review]
Approved, with some details

File: lib/course.js (hunk 2 above)
Comment (FabianBravoA) [reply lib/course.js:263]: maybe just "location"

**Pull request closed without merging**